test/index.js

```diff
@@ -1463,6 +1463,139 @@ describe('Nightmare', function () {
       didSubscribe.should.be.false;
     });
 
+    it('should subscribe to frames when requested necessary', function(done) {
+      var didSubscribe = false;
+      var didUnsubscribe = false;
+      var FrameManager = require('../lib/frame-manager.js');
+      var fn;
+      var manager = FrameManager({
+        webContents: {
+          debugger: {
+            isAttached: function() { return true; },
+            sendCommand: function(command) { if (command === 'DOM.highlightRect') { fn('mock-data'); }}
+          },
+          beginFrameSubscription: function(_fn) { didSubscribe = true; fn = _fn; },
+          endFrameSubscription: function() { didUnsubscribe = true; },
+          executeJavaScript: function() {}
+        }
+      });
+      manager.requestFrame(function (data) {
+        didSubscribe.should.be.true;
+        didUnsubscribe.should.be.true;
+        data.should.equal('mock-data');
+        done();
+      });
+    });
+
+    it('should support multiple concurrent frame subscriptions', function(done) {
+      var subscribeCount = 0;
+      var unsubscribeCount = 0;
+      var FrameManager = require('../lib/frame-manager.js');
+      var fn = null;
+      var assert = require('assert');
+      var async = require('async');
+      var manager = FrameManager({
+        webContents: {
+          debugger: {
+            isAttached: function() { return true; },
+            sendCommand: function(command) {
+              if (command === 'DOM.highlightRect') {
+                setTimeout(function () {
+                  fn('mock-data');
+                }, 100);
+              }
+            }
+          },
+          beginFrameSubscription: function(_fn) { subscribeCount += 1; assert.strictEqual(fn, null); fn = _fn; },
```

Comment:
> On one hand, it's faking out the beginFrameSubscription to prevent it's receiveFrame multiple times (i.e. programmatic). On the other hand, it's verifying that our async.times doesn't invoke the beginFrameSubscription multiple times (i.e. assertion)

Wellllllllll, in the sense that it’s mocking Electron’s behavior, Electron allows `beginFrameSubscription` to be called multiple times—it just replaces the existing subscriber with no checks: https://github.com/electron/libchromiumcontent/blob/2249b37b0c3cb2e7ab4e333fbd3f87356beba55f/patches/guest_view.patch#L33-L36

I do think it’s totally reasonable for the test to be checking that we never call that method unless we’ve never subscribed or have already ended the subscription. In my book, that’s more of a test assertion.

But either way, I don’t think it’s a huge deal.